### ISSUE TITLE
Relationships: change initiative to measurable

### ISSUE THREAD

**Issue by davidwatkins73** [finos/waltz#4812]
Currently this is done via the  `entity_relationship` table but has the following drawbacks

- difficult to write code against as quite flexible
- cannot express multiple relationship kinds between two entities (this is a UI limitation)
- no foreign keys

Tagging with the **removal** flag as some data will need to be migrated out of `entity_relationship`

New table structure is [sketched out in this diagram](https://dbdiagram.io/d/5e97137439d18f5553fda52b)

**Comment by davidwatkins73** [comment]
@rovats - not sure if you have an opinion on this.  It's taking us away from generic tables like EntityReleationship and towards safer strongly typed tables.  However it will increase the number of tables quite a lot and likely increase boilerplate code (though we may be able to abstract some of that).

**Comment by rovats** [comment]
Yes, this will result in too many relationship tables, which may not be ideal.
Although the generic `entity_relationship` table is also not perfect because the ordering of entities is enforced by convention (eg: change initiative is entity_a).

Other products like ServiceNow also use generic tables to store relationships, and relationship type indicates the order of the relationship.

I think a combination of these two approaches will be ideal, similar to what we already do in a lot of other places, like `server_usage`, `software_usage`, `measurable_rating`, where one `entity_kind` is fixed (eg: `server`, `software`, `measurable`) and the related entities can be of any kind.

So, we could have a `change_initiative_entity_relationship` table instead, which can store mesaurable/app_group/app relationships with CIs:

```change_initiative_id | entity_kind | entity_id | rel_kind | description | provenance | last_updated_at | last_updated_by```

A `decorator` table might not be needed here, because a relationship cannot really exist without a `rel_kind`, as opposed to data flows which can exist without data type decorations, and a single table will be much easier to maintain.

Also, each rel_kind might need a different description, even for the same entity, especially true for CI to App relationships.

**Comment by davidwatkins73** [comment]
Thanks for the comments @rovats.  The decorator table is there because there is a strong desire within DB to have multiple relationship kinds between the same two entities.  For example:

`CI-123 -[ Enhancement:Regulatory ,  Enhancement:Efficiency]-> Process-ABC`

Indicates that CI-123 is delivering both regulatory and efficiency enhancements aligned to Process ABC.

**Comment by rovats** [comment]
@davidwatkins73 , agreed about multiple relationships, eg: App can have both functional/non-functional changes delivered under a CI.

I guess a decorator table will make this cleaner, instead of repeating child entity details in the main table.

**Comment by davidwatkins73** [comment]
Think we'll shunt this to the next release as more thought is required.  Though I think we have broad consensus:

- limited new tables (i.e. left hand side of the relationship is of a fixed type, right hand side is free)
- relationship kinds are in a new table and may have optional comments.
- when the last relationship kind is removed then the overall relationship _should_ be removed

**Comment by davidwatkins73** [comment]
...how about something [like this](https://dbdiagram.io/d/5e97137439d18f5553fda52b) ?

Two options:

**First** is revisiting the many tables, but toning down the join tables.  I think jOOQ can minimize boiler plate if we go down this route.  The `relationship_kind` table allows us to restrict relationships on  entity_kind.  At enhancement for the relationship kind table would be to add category restrictions (if one, or both, sides of the reln is a measurable).  We could also look to add read-only flags etc.

**Second** is to have a simpler decorator table (as discussed above) modeled on the logical_flow_decorator approach (one fixed entity, CI, plus  free form entity).

**Comment by rovats** [comment]
Looks like we are going with option 1, a different `relationship_kind` table is good, it also helps us externalise the rel kinds, so they can be customised without a core change.

**Comment by davidwatkins73** [comment]
It's not 100% set in stone yet - but thought we'd try it for this one case.  We can revert to a more generic structure if it's turning out to be too much effort to maintain this approach.   Really would like to get a bit structural integrity in the system and it makes the table design very transparent 
(hopefully)

**Comment by davidwatkins73** [comment]
...and yes - having an externalized `relationship_kind` table should take away one of the main pain points in extending/customizing Waltz.
